pq-crypto/sike_r3/sikep434r3_kem.c

```diff
@@ -0,0 +1,155 @@
+
+    // Encrypt
+    EphemeralKeyGeneration_A(ephemeralsk, ct);
+    EphemeralSecretAgreement_A(ephemeralsk, pk, jinvariant);
+    shake256(h, SIKE_P434_R3_MSG_BYTES, jinvariant, SIKE_P434_R3_FP2_ENCODED_BYTES);
+    for (int i = 0; i < SIKE_P434_R3_MSG_BYTES; i++) {
+        ct[i + SIKE_P434_R3_PUBLIC_KEY_BYTES] = temp[i] ^ h[i];
+    }
+
+    // Generate shared secret ss <- H(m||ct)
+    memcpy(&temp[SIKE_P434_R3_MSG_BYTES], ct, SIKE_P434_R3_CIPHERTEXT_BYTES);
+    shake256(ss, SIKE_P434_R3_SHARED_SECRET_BYTES, temp, SIKE_P434_R3_CIPHERTEXT_BYTES+SIKE_P434_R3_MSG_BYTES);
+
+    // SUCCESS
+    return 1;
+}
+
+// SIKE's decapsulation
+// Input:   secret key sk         (SIKE_P434_R3_SECRET_KEY_BYTES = SIKE_P434_R3_MSG_BYTES + SIKE_P434_R3_SECRETKEY_B_BYTES + SIKE_P434_R3_PUBLIC_KEY_BYTES bytes)
+//          ciphertext message ct (SIKE_P434_R3_CIPHERTEXT_BYTES = SIKE_P434_R3_PUBLIC_KEY_BYTES + SIKE_P434_R3_MSG_BYTES bytes)
+// Outputs: shared secret ss      (SIKE_P434_R3_SHARED_SECRET_BYTES bytes)
+int sike_p434_r3_crypto_kem_dec(unsigned char *ss, const unsigned char *ct, const unsigned char *sk)
+{
+    unsigned char ephemeralsk_[SIKE_P434_R3_SECRETKEY_A_BYTES];
+    unsigned char jinvariant_[SIKE_P434_R3_FP2_ENCODED_BYTES];
+    unsigned char h_[SIKE_P434_R3_MSG_BYTES];
+    unsigned char c0_[SIKE_P434_R3_PUBLIC_KEY_BYTES];
+    unsigned char temp[SIKE_P434_R3_CIPHERTEXT_BYTES+SIKE_P434_R3_MSG_BYTES];
+
+    // Decrypt
+    EphemeralSecretAgreement_B(sk + SIKE_P434_R3_MSG_BYTES, ct, jinvariant_);
+    shake256(h_, SIKE_P434_R3_MSG_BYTES, jinvariant_, SIKE_P434_R3_FP2_ENCODED_BYTES);
+    for (int i = 0; i < SIKE_P434_R3_MSG_BYTES; i++) {
+        temp[i] = ct[i + SIKE_P434_R3_PUBLIC_KEY_BYTES] ^ h_[i];
+    }
+
+    // Generate ephemeralsk_ <- G(m||pk) mod oA
+    memcpy(&temp[SIKE_P434_R3_MSG_BYTES], &sk[SIKE_P434_R3_MSG_BYTES + SIKE_P434_R3_SECRETKEY_B_BYTES], SIKE_P434_R3_PUBLIC_KEY_BYTES);
+    shake256(ephemeralsk_, SIKE_P434_R3_SECRETKEY_A_BYTES, temp, SIKE_P434_R3_PUBLIC_KEY_BYTES+SIKE_P434_R3_MSG_BYTES);
+    ephemeralsk_[SIKE_P434_R3_SECRETKEY_A_BYTES - 1] &= SIKE_P434_R3_MASK_ALICE;
+
+    // Generate shared secret ss <- H(m||ct), or output ss <- H(s||ct) in case of ct verification failure
+    EphemeralKeyGeneration_A(ephemeralsk_, c0_);
+
+    // Verify ciphertext.
+    // If c0_ and ct are NOT equal, decaps failed and we overwrite the shared secret
+    // with pseudorandom noise (ss = H(s||ct)) by performing the copy (dont_copy = false).
+    //
+    // If c0_ and ct are equal, then decaps succeeded and we skip the overwrite and output
+    // the actual shared secret: ss = H(m||ct) (dont_copy = true).
+    if (CRYPTO_memcmp(c0_, ct, SIKE_P434_R3_PUBLIC_KEY_BYTES) != 1) {
+        if (constant_time_copy_or_dont(temp, sk, SIKE_P434_R3_MSG_BYTES, false) != 1) {
+            // FAILURE
+            return 0;
+        }
+    } else {
+        constant_time_copy_or_dont(temp, sk, SIKE_P434_R3_MSG_BYTES, true);
+    }
```

Comment:
This whole if-else clause is needlessly complex, and doesn't take advantage of the functionality of `CRYPTO_memcmp` or `constant_time_copy_or_dont`. 

if `c0_` and `ct` are equal, then `CRYPTO_memcmp` will return 0. If they are not equal, it will return non-zero. Remembering that non-zero` == true` and `0 == false`, you can achieve what you want without any `if` statements (which is the goal!).

If `c0_` and `ct` are equal, then the compare function will return 0, and going by the above comment, we want assign `dont_copy = true`. So this should give us what we want:

```
int dont_copy = !(CRYPTO_memcmp(c0_, ct, SIKE_P434_R3_PUBLIC_KEY_BYTES);
constant_time_copy_or_dont(temp, sk, SIKE_P434_R3_MSG_BYTES, dont_copy);
```

(Double check that my boolean logic is correct here.)